src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java

```diff
@@ -248,34 +247,75 @@ public void run() {
-        List<FoundIndication> foundIndicationList = new LinkedList<FoundIndication>();
-        for (Indication indication : indicationList) {
-            Thread.currentThread().setName("BFA-scanner-"
-                                                   + build.getFullDisplayName() + ": "
-                                                   + cause.getName() + "-"
-                                                   + indication.getUserProvidedExpression());
-            FoundIndication foundIndication = findIndication(indication, build, buildLog);
-            if (foundIndication != null) {
-                foundIndicationList.add(foundIndication);
+    private static List<FoundFailureCause> findIndications(final Collection<FailureCause> causes,
+                                                         Run build,
+                                                         PrintStream buildLog) {
+        List<FoundFailureCause> foundFailureCauses = new ArrayList<>();
+
+        List<FailureCause> singleLineCauses = new ArrayList<>();
+        List<FailureCause> notOnlySingleLineCauses = new ArrayList<>();
+
+        for (FailureCause cause : causes) {
+            boolean atLeast = false;
+            for (Indication indication : cause.getIndications()) {
+                if (indication instanceof MultilineBuildLogIndication) {
+                    atLeast = true;
+                }
+            }
+
+            if (atLeast) {
+                notOnlySingleLineCauses.add(cause);
+            } else {
+                singleLineCauses.add(cause);
             }
         }
-        if (logger.isLoggable(Level.FINER)) {
-            logger.log(Level.FINER, "[BFA] [{0}] [{1}] {2}ms", new Object[]
-                    {build.getFullDisplayName(),
-                            cause.getName(),
-                            String.valueOf(System.currentTimeMillis() - start), });
+
+        try {
+            try (BufferedReader reader = new BufferedReader(build.getLogReader())) {
+                foundFailureCauses.addAll(
+                        FailureReader.scanSingleLinePatterns(
+                                singleLineCauses,
+                                build,
+                                reader,
+                                build.getLogFile().getName()));
+            }
+
+            for (FailureCause cause : notOnlySingleLineCauses) {
+                List<FoundIndication> foundIndications = new ArrayList<>();
+                for (Indication indication : cause.getIndications()) {
+                    long start = System.currentTimeMillis();
+
+                    Thread.currentThread().setName("BFA-scanner-"
+                            + build.getFullDisplayName() + ": "
+                            + cause.getName() + "-"
+                            + indication.getUserProvidedExpression());
+
+                    foundIndications.add(findIndication(indication, build, buildLog));
```

Comment:
findIndication can return null for the multi-line indications (if they don't match), which adds a null indication and causes an NPE down the line.